scripts/include/helpers.inc.php

```diff
@@ -121,45 +121,98 @@ function mergeXliffDomToAtom(DOMDocument $domA, DOMDocument $domB): DOMDocument
+    }
+  }
+
+  // Elements in A AND B -> check source string and if different -> replace in B
+  $transUnitsA = $domA->getElementsByTagName('trans-unit');
+  $transUnitsB = $domB->getElementsByTagName('trans-unit');
+
+  $i = 0;
+  foreach ($transUnitsA as $transUnit) {
+    $idListA[] = $transUnitsA->item($i)->getAttribute('id');
+    $i++;
+  }
 
-  foreach ($sourcesA as $source) {
-    $stringsA[] = $source->nodeValue;
+  $i = 0;
+  foreach ($transUnitsB as $transUnit) {
+    $idListB[] = $transUnitsB->item($i)->getAttribute('id');
+    $i++;
   }
 
-  foreach ($sourcesB as $source) {
-    $stringsB[] = $source->nodeValue;
+  // Items in A that are also in B.
+  $verifyInB = array_intersect($idListA, $idListB);
+
+  // Use trans-unit's id attribute to identify strings. When one is
+  // matched, should also compare the source string target and if
+  // different, replace it in target.
+  $i = 0;
+  foreach ($transUnitsA as $transUnit) {
+    $id = $transUnitsA->item($i)->getAttribute('id');
+
+    if (in_array($id, $verifyInB)) {
+      $node = $domB->getElementById($id);
+
+      // Get 'source' string element from dom B trans-unit.
+      $sourcesB = $node->getElementsByTagName('source');
+      foreach ($sourcesB as $source) {
+        $stringsB[] = $source->nodeValue;
+      }
+      $stringB = implode("", $stringsB);
+
+      $sourcesA = $transUnitsA->item($i)->getElementsByTagName('source');
+      foreach ($sourcesA as $source) {
+        $stringsA[] = $source->nodeValue;
+      }
+      $stringA = implode("", $stringsA);
+
+      // Compare strings. If different, delete and replace trans-unit.
+      if ($stringA !== $stringB) {
+        // Remove trans-unit as child of domB body
+        $bodyB = $domB->getElementsByTagName('body')[0];
+        $bodyB->removeChild($node);
+
+        // Add matching trans-unit from domA.
+        $bodyB->appendChild($domB->importNode($transUnitsA->item($i), true));
+      }
+    }
+
+    $i++;
   }
 
-  $copyToB = array_diff($stringsA, $stringsB);
+  // Elements in A but not in B -> copy to B
+  $copyToB = array_diff($idListA, $idListB);
 
-  foreach($sourcesA as $source) {
-    if (in_array($source->nodeValue, $copyToB)) {
-      // get parent which should be a trans-unit
-      $parent = $source->parentNode;
-      // add trans-unit as child of domB body
+  foreach ($copyToB as $id) {
+    $node = $domA->getElementById($id);
+
+    if (null !== $node) {
       $bodyB = $domB->getElementsByTagName('body')[0];
-      $bodyB->appendChild($domB->importNode( $parent, true ));
+
+      // Add matching trans-unit from domA.
+      $bodyB->appendChild($domB->importNode($node, true));
     }
   }
 
   // For elements in B not in A -> delete from B
-  // Refresh string array
-  foreach ($sourcesB as $source) {
-    $stringsB[] = $source->nodeValue;
+  // Refresh id list values for domB as we modified them above.
+  $i = 0;
+  $idListB = [];
+  foreach ($transUnitsB as $transUnit) {
```

Comment:
I think `$transUnitsB` is stale here too, because you are not directly modifying it above?

Reply:
I have removed this update of the id list as the update to B is inconsequential. The items that were copied into B before this should be ignored because this check is for items that are in B but not in A - this cannot happen with the items copied from A -> B as occurs above.